### PR TITLE
67782 - Reduce Topic, Queue and Subscription enumerations

### PR DESCRIPTION
**Pull request by devr24**
- added extension methods for retrieving entity by name      
- added test cases for the three new extension methods
- reduce enumerations in CreateIfNotExists methods (topic, topic-subscription, queue)

### CODE REVIEW

**Comment by devr24** [comment]
Reduce the number of enumerations on Topics, Queues and Subscriptions

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs [hunk 1]
```diff
@@ -34,7 +34,7 @@ await sbNamespace.Queues
                              .CreateAsync();
 
             await sbNamespace.RefreshAsync();
```
Comment (VaclavK) [inline src/Eshopworld.Messaging/ServiceBusExtensions.cs]: so this should not be necessary anymore then if you are not enumerating to locate the topic/queue etc..

so to be explicit, this likely applies to those other artefact methods below

me thinks

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs [hunk 2]
```diff
@@ -22,10 +22,10 @@ public static class ServiceBusNamespaceExtensions
         /// <returns>The <see cref="IQueue"/> entity object that references the Azure queue.</returns>
         public static async Task<IQueue> CreateQueueIfNotExists(this IServiceBusNamespace sbNamespace, string name)
         {
-            var queue = (await sbNamespace.Queues.ListAsync()).SingleOrDefault(q => q.Name == name.ToLower());
+            var queue = await sbNamespace.GetQueueByName(name);
```
Comment (dguerin) [inline src/Eshopworld.Messaging/ServiceBusExtensions.cs]: i think we need to enforce lowercase naming here as well. ie: name.ToLower()

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 1 above)
Comment (devr24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: Hey Oisin,
Just to confirm - I can go ahead and remove the `await sbNamespace.RefreshAsync();`? as this has been made redundant?
Do you think the `RefreshAsync` should be moved to the new extension methods I have created?
Thanks

Comment (VaclavK) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: will call you in a bit

Comment (devr24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: I can confirm that the refresh is not necessary.  The namespace is up to date following the create call.  Oisin noticed this in the SDK samples, and my tests confirm it (there's a get entity before create, then create entity, then get entity after, all of which work as expected).

**Review by micszym — APPROVED** [review]

**Review by VaclavK — APPROVED** [review]
lgtm

@rmccabe24 the only other comment I would make - and not caused by you by this PR - is that we may have a race condition here... if the existence check fails and something creates it before the create call, I am not sure if create will handle that.... or we shoudl catch and refresh and return "by name"

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs [hunk 3]
```diff
@@ -88,6 +85,69 @@ public static string GetNamespaceNameFromConnectionString(this string connection
...
         }
+
+        /// <summary>
+        /// Gets a topic by name, will return null if not found.
+        /// More efficient than enumerating all topics and find by name.
+        /// </summary>
+        /// <param name="sbNamespace">The sb namespace to extend.</param>
+        /// <param name="name">The name of the topic to find.</param>
+        /// <returns>The <see cref="ITopic"/> if exists and null if not.</returns>
+        public static async Task<ITopic> GetTopicByName(this IServiceBusNamespace sbNamespace, string name)
```
Comment (dguerin) [inline src/Eshopworld.Messaging/ServiceBusExtensions.cs]: This was a convention at the time but I believe all methods which return a Task should also be postfixed with Async. It just becomes clearer to the developer and follows industry convention.

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs [hunk 4]
```diff
@@ -88,6 +85,69 @@ public static string GetNamespaceNameFromConnectionString(this string connection
...
+        }
+
+        /// <summary>
+        /// Gets a queue by name, will return null if not found.
+        /// More efficient than enumerating all queues and finding by name.
+        /// </summary>
+        /// <param name="sbNamespace">The sb namespace to extend.</param>
+        /// <param name="name">The name of the topic to find.</param>
+        /// <returns>The <see cref="ITopic"/> if exists and null if not.</returns>
+        public static async Task<IQueue> GetQueueByName(this IServiceBusNamespace sbNamespace, string name)
```
Comment (dguerin) [inline src/Eshopworld.Messaging/ServiceBusExtensions.cs]: Ditto on names

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 2 above)
Comment (VaclavK) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: that is done inside the GetQueueByName

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs [hunk 5]
```diff
@@ -88,6 +85,69 @@ public static string GetNamespaceNameFromConnectionString(this string connection
...
+        }
+
+        /// <summary>
+        /// Gets a topic subscription by name, will return null if not found.
+        /// More efficient than enumerating all subscriptions and finding by name.
+        /// </summary>
+        /// <param name="topic">The topic to extend.</param>
+        /// <param name="name">The name of the subscription to find.</param>
+        /// <returns>The <see cref="ITopic"/> if exists and null if not.</returns>
+        public static async Task<ISubscription> GetSubscriptionByName(this ITopic topic, string name)
```
Comment (dguerin) [inline src/Eshopworld.Messaging/ServiceBusExtensions.cs]: ditto

**Comment by dguerin** [comment]
> lgtm
> 
> @rmccabe24 the only other comment I would make - and not caused by you by this PR - is that we may have a race condition here... if the existence check fails and something creates it before the create call, I am not sure if create will handle that.... or we shoudl catch and refresh and return "by name"

Rob and I talked about that this morning as well but really that may only happen if n number of devs or the tests are run from a ci/cd pipeline at the same time. Which as far as i know has not happened. If it does happen once then I think it's something we should look at.

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 3 above)
Comment (VaclavK) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: I remember conversation a long time ago and the opposite decision

async-ness is made clear by the return type so the suffix is additive and you are also warned explicitly if you do not await

I don't mind either just wanted to highlight what was decided then

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 2 above)
Comment (dguerin) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: this was something I had written this morning when Rob and I were talking so it was subsequently fixed. I posted all the reviews together now.

**Comment by VaclavK** [comment]
> 
> 
> > lgtm
> > @rmccabe24 the only other comment I would make - and not caused by you by this PR - is that we may have a race condition here... if the existence check fails and something creates it before the create call, I am not sure if create will handle that.... or we shoudl catch and refresh and return "by name"
> 
> Rob and I talked about that this morning as well but really that may only happen if n number of devs or the tests are run from a ci/cd pipeline at the same time. Which as far as i know has not happened. If it does happen once then I think it's something we should look at.

I think this goes beyond the tests

this is usage.. if you have multiple instances of the service and each running these routines...

**Comment by dguerin** [comment]
@rmccabe24 don't forget to bump the version number and add a description in the .csproj Patch bump would be good with me.

**Comment by VaclavK** [comment]
> 
> 
> @rmccabe24 don't forget to bump the version number and add a description in the .csproj Patch bump would be good with me.

good catch...

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 3 above)
Comment (tabman83) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: It's MS practice to have -Async postfix for async method names, and it improves readability imho. +1 from me

**Review by tabman83 — APPROVED** [review]

**Comment by VaclavK** [comment]
@tabman83 good explanation IMHO why the async suffix was introduced initially (it is effectively a legacy artefact originally but maybe a good convention on its own  )... 

https://docs.particular.net/nservicebus/upgrades/5to6/async-suffix

from me 
+1
-1

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 2 above)
Comment (devr24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: Yes guys, I added this into the GetByName extensions following David's initial review.  Good to have visibility of that here.  Thanks guys.

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 3 above)
Comment (devr24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: This one's a bit subjective...

Async became a convention when async/await was introduced, to differentiate the return type (usually came with a sync and async version of a method that couldn't have the same name).  

IMO, I think these days this is inferred by the async Task being returned and it's more closely sticking with Hungarian notation (potentially we could use if we had a sync and async version of a method).  However, I don't know eshops overall stance on it.  

When I wrote this code, I did scan the rest of the package for use of Async as a suffix being used in any other methods and it wasn't used anywhere else - so I was happy to not include.  

I'm happy to change either way guys, let me know what you think.

Comment (rmccabe24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs:96]: I'm happy to change either way guys, let me know what you think.

**Comment by devr24** [comment]
> > > lgtm
> > > @rmccabe24 the only other comment I would make - and not caused by you by this PR - is that we may have a race condition here... if the existence check fails and something creates it before the create call, I am not sure if create will handle that.... or we shoudl catch and refresh and return "by name"
> > 
> > 
> > Rob and I talked about that this morning as well but really that may only happen if n number of devs or the tests are run from a ci/cd pipeline at the same time. Which as far as i know has not happened. If it does happen once then I think it's something we should look at.
> 
> I think this goes beyond the tests
> 
> this is usage.. if you have multiple instances of the service and each running these routines...

As you mentioned, the code I've introduced hasn't changed the flow, so this race condition (to create entity) would have been in the existing code.

That being said, just thinking about this possible race condition - if two clients made calls to CreateIfNotExists simultaneously, wouldn't one be successful and the subsequent call find that it exists and just return the newly created entity anyway?  I'll test this guys and get back to you on it.

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 3 above)
Comment (VaclavK) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: I was reflecting on this many times for my own stuff and one argument I come to myself (besides the whole return type and squiiqly highlight if I forget await) is that for this kind of code you are dealing with async by default in essence...

I would even turn it around these days.. if not async. put suffix :D

**Comment by VaclavK** [comment]
> 
> 
> > > > lgtm
> > > > @rmccabe24 the only other comment I would make - and not caused by you by this PR - is that we may have a race condition here... if the existence check fails and something creates it before the create call, I am not sure if create will handle that.... or we shoudl catch and refresh and return "by name"
> > > 
> > > 
> > > Rob and I talked about that this morning as well but really that may only happen if n number of devs or the tests are run from a ci/cd pipeline at the same time. Which as far as i know has not happened. If it does happen once then I think it's something we should look at.
> > 
> > 
> > I think this goes beyond the tests
> > this is usage.. if you have multiple instances of the service and each running these routines...
> 
> Wasn't the refresh done after the create though, so this potential race condition could have already existed? Both the ListTopicAsync and GetTopicByName (just to use one entity type as an example) both make back end calls to check this. This fix was just to stop the longer enumeration call of ListTopicsAsync (in this example) - not sure this changes that flow.
> 
> Although thinking about that race condition, if two clients made calls to create simultaneously, wouldn't one be successful and the subsequent call find that it exists and just return the newly created entity anyway? I'll test this.

cheers....should be easy to test even with just one client.. put a breakpoint and create an entity by porta or SB explorer or etc... and f5

**Comment by devr24** [comment]
> > > > > lgtm
> > > > > @rmccabe24 the only other comment I would make - and not caused by you by this PR - is that we may have a race condition here... if the existence check fails and something creates it before the create call, I am not sure if create will handle that.... or we shoudl catch and refresh and return "by name"
> > > > 
> > > > 
> > > > Rob and I talked about that this morning as well but really that may only happen if n number of devs or the tests are run from a ci/cd pipeline at the same time. Which as far as i know has not happened. If it does happen once then I think it's something we should look at.
> > > 
> > > 
> > > I think this goes beyond the tests
> > > this is usage.. if you have multiple instances of the service and each running these routines...
> > 
> > 
> > Wasn't the refresh done after the create though, so this potential race condition could have already existed? Both the ListTopicAsync and GetTopicByName (just to use one entity type as an example) both make back end calls to check this. This fix was just to stop the longer enumeration call of ListTopicsAsync (in this example) - not sure this changes that flow.
> > Although thinking about that race condition, if two clients made calls to create simultaneously, wouldn't one be successful and the subsequent call find that it exists and just return the newly created entity anyway? I'll test this.
> 
> cheers....should be easy to test even with just one client.. put a breakpoint and create an entity by porta or SB explorer or etc... and f5

---

I looked into this and as expected, if the race condition happens and another process creates the same entity, then an error response is returned.  The response is returned as BadRequest.  Unfortunately, it only gives details of the BadRequest in the message string - list of possible responses:  https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-resource-manager-exceptions

Here's the full response body from the `CloudException returned when there is a conflict of this type::

```json
{"error":{"message":"SubCode=40000. The value for the requires duplicate detection property of an existing Topic cannot be changed. To know more visit https://aka.ms/sbResourceMgrExceptions. TrackingId:3f2ba99f-b3bb-4f8d-8737-f635ec9ab715_M6SN1_M6SN1_G11, SystemTracker:cloud1core.servicebus.windows.net:testgettopicbyname, Timestamp:2020-05-14T23:01:02","code":"BadRequest"}}
```
Taking topic as an example, the CreateIfNotExists method would potentially turn into:

```csharp
public static async Task<ITopic> CreateTopicIfNotExists(this IServiceBusNamespace sbNamespace, string name)
{
    var topic = await sbNamespace.GetTopicByName(name);
    if (topic != null) return topic;

    try
    {
        return await sbNamespace.Topics
                     .Define(name)
                     .WithDuplicateMessageDetection(TimeSpan.FromMinutes(10))
                     .CreateAsync();
    }
    catch (CloudException ce)
        when (ce.Response.StatusCode == HttpStatusCode.BadRequest && 
              ce.Message.Contains("SubCode=40000. The value for the requires duplicate detection property of an existing Topic cannot be changed"))
    {
        // Creation race condition occurred. Return existing topic.
        return await sbNamespace.GetTopicByName(name);
    }
}
```

We need to recall the GetTopicByName to return the topic created by the other process.  An extra call but probably fine as a fallback - this is a very rare condition but one we potentially should handle.  It's a bit messy though.  Should the caller handle the CloudException?  Let me know your thoughts guys and I can make this further amendment.

**Comment by devr24** [comment]
@VaclavK @dguerin 
I also checked whether the .ToLower was required for both the create and the getByNameAsync call.  Neither need it - it must normalise the name internally and always uses the lowercase name.

Me passing an entity name of "ROBTEST" both created a topic as "robtest" and then found the same topic when looking up using the uppercase version.

With this in mind, should we drop the .toLower within these methods as its unnecessary?

**Comment by VaclavK** [comment]
> 
> I looked into this and as expected, if the race condition happens and another process creates the same entity, then an error response is returned. The response is returned as BadRequest. Unfortunately, it only gives details of the BadRequest in the message string - list of possible responses: https://docs.microsoft.com/en-us/azure/service-bus-messaging/service-bus-resource-manager-exceptions
> 
> Here's the full response body from the `CloudException returned when there is a conflict of this type::
> 
> ```json
> {"error":{"message":"SubCode=40000. The value for the requires duplicate detection property of an existing Topic cannot be changed. To know more visit https://aka.ms/sbResourceMgrExceptions. TrackingId:3f2ba99f-b3bb-4f8d-8737-f635ec9ab715_M6SN1_M6SN1_G11, SystemTracker:cloud1core.servicebus.windows.net:testgettopicbyname, Timestamp:2020-05-14T23:01:02","code":"BadRequest"}}
> ```
> 
> Taking topic as an example, the CreateIfNotExists method would potentially turn into:
> 
> ```cs
> public static async Task<ITopic> CreateTopicIfNotExists(this IServiceBusNamespace sbNamespace, string name)
> {
>     var topic = await sbNamespace.GetTopicByName(name);
>     if (topic != null) return topic;
> 
>     try
>     {
>         return await sbNamespace.Topics
>                      .Define(name)
>                      .WithDuplicateMessageDetection(TimeSpan.FromMinutes(10))
>                      .CreateAsync();
>     }
>     catch (CloudException ce)
>         when (ce.Response.StatusCode == HttpStatusCode.BadRequest && 
>               ce.Message.Contains("SubCode=40000. The value for the requires duplicate detection property of an existing Topic cannot be changed"))
>     {
>         // Creation race condition occurred. Return existing topic.
>         return await sbNamespace.GetTopicByName(name);
>     }
> }
> ```
> 
> We need to recall the GetTopicByName to return the topic created by the other process. An extra call but probably fine as a fallback - this is a very rare condition but one we potentially should handle. It's a bit messy though. Should the caller handle the CloudException? Let me know your thoughts guys and I can make this further amendment.

these findings are great...

my 2cents.. looks to me as graceful handling..

if we do return exceptions, the client would effectively have to run the same code

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 3 above)
Comment (dguerin) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: yes to all the above as you define. I quite like David Fowlers opinion on it in that most things are async but when it's becomes muddled you need to define. However, it's not always clear that a method can be async if the return type is void, which can happen with timers or button click events when you're on a WPF for example. 

In cases like this it makes sense to be explicit. Not everyone has resharper installed to catch this and general understood convention works better. One bug I saw last week was related to this very problem, an async was not handed with await.

Comment (dguerin) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: Platform are the owners of this code so they should have the deciding vote.

**Comment by dguerin** [comment]
> > > > lgtm
> > > > @rmccabe24 the only other comment I would make - and not caused by you by this PR - is that we may have a race condition here... if the existence check fails and something creates it before the create call, I am not sure if create will handle that.... or we shoudl catch and refresh and return "by name"
> > > 
> > > 
> > > Rob and I talked about that this morning as well but really that may only happen if n number of devs or the tests are run from a ci/cd pipeline at the same time. Which as far as i know has not happened. If it does happen once then I think it's something we should look at.
> > 
> > 
> > I think this goes beyond the tests
> > this is usage.. if you have multiple instances of the service and each running these routines...
> 
> As you mentioned, the code I've introduced hasn't changed the flow, so this race condition (to create entity) would have been in the existing code.
> 
> That being said, just thinking about this possible race condition - if two clients made calls to CreateIfNotExists simultaneously, wouldn't one be successful and the subsequent call find that it exists and just return the newly created entity anyway? I'll test this guys and get back to you on it.

Super edge case that I think with eda would almost never happen because of the partitioned naming convention that is in place.

**Comment by dguerin** [comment]
> @VaclavK @dguerin
> I also checked whether the .ToLower was required for both the create and the getByNameAsync call. Neither need it - it must normalise the name internally and always uses the lowercase name.
> 
> Me passing an entity name of "ROBTEST" both created a topic as "robtest" and then found the same topic when looking up using the uppercase version.
> 
> With this in mind, should we drop the .toLower within these methods as its unnecessary?

if it's not needed on the new stuff, I'd leave it out.

**Comment by dguerin** [comment]
@rmccabe24 don't forget to add your name to the author list :)

**Review by dguerin — APPROVED** [review]

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 3 above)
Comment (devr24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: @tabman83 - will I add the Async suffix? :-) I dont mind at all, so let me know sure.

Comment (tabman83) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: It's very thin line. I'd rather have the async suffix as it adds to clarity imho and is also what I have been used for years now. However I haven't come across a widespread usage of the async suffix in the eshop code I've worked so far. Given this I would prefer consistency over clarity. Maybe @dguerin or @VaclavK who have a broader visibility of the code base can advise what is the most used convention across.

Comment (VaclavK) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: to move this forward my suggestion is to add the suffix

this does not introduce any debt, may prevent futher conversations  and may increase clarity so seems win-win situation,..

Comment (devr24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: Sounds good - I'll push that now

Comment (devr24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: This is fixed in the latest commit guys

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 4 above)
Comment (devr24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: Fixed now in latest commit

File: src/Eshopworld.Messaging/ServiceBusExtensions.cs (hunk 5 above)
Comment (devr24) [reply src/Eshopworld.Messaging/ServiceBusExtensions.cs]: Fixed now in latest commit
